File: frontend/src/app/domain/commons/result.ts
import { BaseError } from "src/app/_services/_models/base.model";

export class Result {
    succeeded: boolean;
    errors: BaseError[] = [];
    constructor(succeeded: boolean, errors: BaseError[]) {
        this.succeeded = succeeded;
        this.errors = errors;
    }

    static Success(): Result {
        return new Result(true, []);
    }

    static Failure(errors: BaseError[]): Result {
        return new Result(false, errors);
    }
}

export class DataResult<T> {
    data: T = undefined;
    succeeded: boolean;
    errors: BaseError[] = [];

    constructor(data: T, succeeded: boolean, errors: BaseError[]) {
        this.data = data;
        this.succeeded = succeeded;
        this.errors = errors;
    }

    Success(data: T): DataResult<T> {
        return new DataResult(data, true, []);
    }

    Failure(errors: BaseError[]): DataResult<T> {
        return new DataResult(undefined, false, errors);
    }
}

File: backend/src/Domain/Common/Results/Result.cs
namespace Domain.Common.Results
{
    public class Result
    {
        protected internal Result(bool succeeded, List<Error> errors)
        {
            Succeeded = succeeded;
            Errors = errors;
        }

        public bool Succeeded { get; init; }

        public List<Error> Errors { get; set; }

        public static Result Success() => new Result(true, []);
        public static Result Failure(List<Error> errors) => new Result(false, errors);
        public static Result Failure(Error error) => new Result(false, [error]);

        public static Result<TData> Success<TData>(TData data) => new(data, true, []);
        public static Result<TData> Failure<TData>(Error error) => new(default, false, [error]);
        public static Result<TData> Failure<TData>(List<Error> errors) => new(default, false, errors);
    }
    public class Result<TData> : Result
    {
        private readonly TData? _data;
        protected internal Result(TData? data, bool succeeded, List<Error> errors) : base(succeeded, errors)
        {
            _data = data;
        }

        public TData Data => Succeeded
            ? _data!
            : throw new InvalidOperationException("The data of the failure result can't be accessed.");

        public static implicit operator Result<TData>(TData? data)
            => data is not null ? Success(data) : Failure<TData>(Error.NullData);

    }
}
